### PR TITLE
Add schools import script

### PR DESCRIPTION
**Pull request by tvararu**
This will be executed on new environments in order to populate the database with the full list of schools based on the Get Information about Schools service from the DFE.

Also adds the GIAS CSV to version control which is 62MB large and over the 50MB warning limit that GitHub issues. Their suggestion is to use `git-lfs` for managing these sort of files, but since they don't outright block files under 100MB, I think it's simpler to not use LFS right this moment since the CSV is unlikely to double in size any time soon.

TODO:

- [x] Add it to the seeds script so it runs on new environments and Heroku
- [x] Use https://github.com/zdennis/activerecord-import for potentially much faster imports
- [x] Add download script
- [x] Use gzip to reduce file size

#### Without activerecord-import

```bash
mavis (gias-schools) $ rails db:seed:replant
mavis (gias-schools) $ time rails schools:import
Starting schools import. Total locations: 1
Progress: 51128/51128 [##################################################] 100.0%
Schools import completed. Total locations: 51128

________________________________________________________
Executed in  185.26 secs    fish           external
   usr time  124.03 secs    0.07 millis  124.03 secs
   sys time    6.69 secs    1.79 millis    6.69 secs
mavis (gias-schools) $ time rails schools:import
Starting schools import. Total locations: 51128
Progress: 51128/51128 [##################################################] 100.0%
Schools import completed. Total locations: 51128

________________________________________________________
Executed in   35.60 secs    fish           external
   usr time   29.13 secs    0.08 millis   29.13 secs
   sys time    1.53 secs    1.94 millis    1.53 secs
```

#### With activerecord-import

```bash
mavis (gias-schools) $ rails db:seed:replant
mavis (gias-schools) $ time rails schools:import
Starting schools import. Total locations: 1
Progress: 51128/51128 [##################################################] 100.0%
Schools import completed. Total locations: 51128

________________________________________________________
Executed in   16.91 secs    fish           external
   usr time   14.15 secs    0.09 millis   14.15 secs
   sys time    0.44 secs    1.79 millis    0.44 secs

mavis (gias-schools) $ time rails schools:import
Starting schools import. Total locations: 51128
Progress: 51128/51128 [##################################################] 100.0%
Schools import completed. Total locations: 51128

________________________________________________________
Executed in   17.29 secs    fish           external
   usr time   14.18 secs    0.10 millis   14.18 secs
   sys time    0.45 secs    1.05 millis    0.45 secs
```

### CODE REVIEW

**Comment by tvararu** [comment]
```bash
$ heroku run bash -a mavis-pr-1586
Running bash on ⬢ mavis-pr-1586... up, run.3671 (Basic)
~ $ bin/rails console
Loading production environment (Rails 7.1.3.4)
irb(main):001> Location.count
=> 51128
```

**Comment by tvararu** [comment]
I attempted to write a download script, and got this far, but it fails and I'm not sure it's worth using. Leaving here for posterity:

```ruby
  # To get the latest version of the zip:
  # 1. Go to https://get-information-schools.service.gov.uk/Downloads
  # 2. Check "Establishment fields CSV"
  # 3. Submit
  # 4. Download the zip file
  # 5. Place it in db/data/edubasealldata.zip
  #
  # Alternatively, you can run this task.
  desc "Download schools data"
  task download: :environment do
    require "capybara/cuprite"

    puts "Starting schools data download process..."
    puts "Visiting the downloads page"
    browser = Capybara::Session.new(:cuprite)
    browser.visit "https://get-information-schools.service.gov.uk/Downloads"

    puts "Checking the establishment fields CSV checkbox"
    browser.find("#establishment-fields-csv-checkbox", visible: false).check

    puts "Clicking the submit button"
    browser.click_button "Download selected files"

    puts "Waiting for the 'Results.zip' button to appear..."
    browser.find_button("Results.zip", wait: 20)
    puts "'Results.zip' button found"

    puts "Clicking the 'Results.zip' button to download the file..."
    browser.click_button("Results.zip")

    puts "Waiting for the file to download..."
    # Wait for the file to be downloaded (adjust the timeout as needed)
    downloaded_file =
      Dir.glob("#{Capybara.save_path}/*.zip").max_by { |f| File.mtime(f) }
    wait_time = 0
    until downloaded_file && File.size(downloaded_file).positive?
      sleep 1
      wait_time += 1
      break if wait_time > 60 # Timeout after 60 seconds
      downloaded_file =
        Dir.glob("#{Capybara.save_path}/*.zip").max_by { |f| File.mtime(f) }
    end

    if downloaded_file
      puts "File downloaded successfully"
      FileUtils.mv(downloaded_file, "db/data/edubasealldata.zip")
      puts "File moved to db/data/edubasealldata.zip"
    else
      puts "Failed to download the file"
    end

    puts "Schools data downloaded to db/data/edubasealldata.zip"
    puts "Download process completed"
  end
```

**Comment by misaka** [comment]
I had a look at the download thing, it looks like Capybara wasn't saving the file anywhere? Mechanize seems to go a better job, I added `mechanize` to `Rubygems` and then converted that script to use it:

```ruby
# To get the latest version of the zip:
# 1. Go to https://get-information-schools.service.gov.uk/Downloads
# 2. Check "Establishment fields CSV"
# 3. Submit
# 4. Download the zip file
# 5. Place it in db/data/edubasealldata.zip
#
# Alternatively, you can run this task.
desc "Download schools data"
task "gias:download": :environment do
  require "mechanize"

  puts "Starting schools data download process..."
  agent = Mechanize.new

  puts "Visiting the downloads page"
  page = agent.get("https://get-information-schools.service.gov.uk/Downloads")

  puts "Checking the establishment fields CSV checkbox"
  form = page.form_with(action: "/Downloads/Collate")
  form.checkbox_with(id: "establishment-fields-csv-checkbox").check

  puts "Submitting the form"
  download_page = form.submit

  # There is a meta refresh on the download_page. Mechanize didn't seem to
  # follow it so we're just refreshing that page manually until the button shows
  # up below.
  wait_time = 0
  until (
          download_form =
            download_page.form_with(action: "/Downloads/Download/Extract")
        ) || wait_time > 60
    puts "Waiting for the 'Results.zip' link to appear..."
    sleep(2)
    wait_time += 2
    download_page_uri = download_page.uri
    download_page = agent.get(download_page_uri)
  end

  if download_form
    download_button = download_form.button_with(value: "Results.zip")
    puts "'Results.zip' link found, downloading the file..."
    download_file = agent.click(download_button)
    download_file.save
    puts "File downloaded successfully to #{download_file.filename}"
  else
    puts "Download button never appeared, abotting"
  end
end
```

And it seems to work:

```shell
$ rails gias:download
Starting schools data download process...
Visiting the downloads page
Checking the establishment fields CSV checkbox
Submitting the form
Waiting for the 'Results.zip' link to appear...
Waiting for the 'Results.zip' link to appear...
'Results.zip' link found, downloading the file...
File downloaded successfully to extract.zip
$ unzip -v extract.zip
Archive:  extract.zip
 Length   Method    Size  Cmpr    Date    Time   CRC-32   Name
--------  ------  ------- ---- ---------- ----- --------  ----
64588771  Defl:N  8004870  88% 08-13-2024 13:07 a2311d70  edubasealldata20240813.csv
--------          -------  ---                            -------
64588771          8004870  88%                            1 file
```

File: lib/tasks/schools_import.rake [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+        if locations.size >= batch_size
+          Location.import locations,
+                          on_duplicate_key_update: {
+                            conflict_target: [:urn],
+                            columns: %i[name address town county postcode url]
+                          }
+          locations.clear
+        end
+
+        # Update progress bar
```
Comment (misaka) [inline lib/tasks/schools_import.rake:56]: `ruby-progressbar` is already in `Gemfile.lock`, what about using that? Also looks easy to use and has 1.6K ⭐s on GitHub!!!

https://github.com/jfelchner/ruby-progressbar

**Review by thomasleese — APPROVED** [review]
I wonder if it's worth creating a `schools:download_and_import` task that does both for you automatically.

**Comment by tvararu** [comment]
@thomasleese we could add that later 👍

**Pull request merged**